### PR TITLE
CVS-46595 Add custom node library manager

### PR DESCRIPTION
**Pull request by dkalinowski**
https://jira.devtools.intel.com/browse/CVS-46595

### CODE REVIEW

**Comment by atobiszei** [comment]
Make sure to use 2021 in copyright headers.

File: src/BUILD [hunk 1]
```diff
@@ -20,9 +20,12 @@ cc_library(
     srcs = [
         "config.cpp",
         "config.hpp",
+        "custom_node_interface.hpp",
+        "custom_node_library_manager.cpp",
+        "custom_node_library_manager.hpp",
         "customloaderconfig.hpp",
-	"customloaders.hpp",
-	"customloaders.cpp",
+	    "customloaders.hpp",
```
Comment (atobiszei) [inline src/BUILD:27]: different type of padding?

Comment (dkalinowski) [reply src/BUILD:27]: This corrects customloaders padding to match other paddings.

File: src/custom_node_library_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    error = dlerror();
+    if (error || releaseBuffer == nullptr) {
+        SPDLOG_LOGGER_INFO(modelmanager_logger, "Library name: {} with error: {}", name, error);
+        dlclose(handle);
+        return StatusCode::NODE_LIBRARY_LOAD_FAILED;
+    }
+
+    release_fn releaseTensors = (release_fn)dlsym(handle, "releaseTensors");
+    error = dlerror();
+    if (error || releaseBuffer == nullptr) {
```
Comment (atobiszei) [inline src/custom_node_library_manager.cpp:59]: Shouldn;t we return different kind of error for missing different part of library?

File: src/node_library.cpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include "node_library.hpp"
+
+namespace ovms {
+
+}  // namespace ovms
```
Comment (atobiszei) [inline src/node_library.cpp:20]: Not required file.

File: src/test/custom_nodes/node_mock.cpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+extern "C" int execute(const struct CustomNodeTensor* inputs, int inputsLength, struct CustomNodeTensor** outputs, int* outputsLength, const struct CustomNodeParam* params, int paramsLength) {
+    return 1;
+}
+
+extern "C" int releaseBuffer(struct CustomNodeTensor* output) {
+    return 2;
+}
+
+extern "C" int releaseTensors(struct CustomNodeTensor* outputs) {
+    return 3;
```
Comment (atobiszei) [inline src/test/custom_nodes/node_mock.cpp:28]: Not sure if for testing purposes it wouldn't be better to return 0, and only construct special node mock that returns errors.

File: src/test/node_library_manager_test.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    EXPECT_EQ(library.execute(nullptr, 0, nullptr, nullptr, nullptr, 0), 1);
+    EXPECT_EQ(library.releaseBuffer(nullptr), 2);
+    EXPECT_EQ(library.releaseTensors(nullptr), 3);
+}
+
+TEST(NodeLibraryManagerTest, LibraryLoadingDuplicateName) {
+    CustomNodeLibraryManager manager;
+    auto status = manager.loadLibrary("random_name", "/ovms/bazel-bin/src/lib_node_mock.so");
+    ASSERT_EQ(status, StatusCode::OK);
+    status = manager.loadLibrary("random_name", "/ovms/bazel-bin/src/lib_node_mock.so");
```
Comment (atobiszei) [inline src/test/node_library_manager_test.cpp:48]: I would differentiate two tests here -> in one case we have the same name, in other the same path. First would be for error handling, the second one for documenting what is expected behavior.

File: src/custom_node_interface.hpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    unsigned char* data;
+    int dataLength;
+    int* dims;
+    int dimsLength;
+    int precision;
+};
+
+struct CustomNodeParam {
+    const char *key, *value;
+};
```
Comment (atobiszei) [inline src/custom_node_interface.hpp:29]: Shouldn't be everything extern C?

**Review by atobiszei — CHANGES_REQUESTED** [review]

**Review by Ficol — APPROVED** [review]

File: src/custom_node_library_manager.cpp (hunk 2 above)
Comment (dkalinowski) [reply src/custom_node_library_manager.cpp:59]: @atobiszei I don't think so. We will never use this information in application. We will just skip library if any error appears. Therefore I think separate log for each part of library missing is enough.

File: src/test/custom_nodes/node_mock.cpp (hunk 4 above)
Comment (dkalinowski) [reply src/test/custom_nodes/node_mock.cpp:28]: I wanted to differentiate return values to be able to identify these functions in unit test. For integration tests and error handling test purposes we can create separate example library, not a scope of this task.

File: src/custom_node_interface.hpp (hunk 6 above)
Comment (dkalinowski) [reply src/custom_node_interface.hpp:29]: `extern "C"` disabled symbol name mangling. From my understanding, we shouldn't care about struct name mangling, all wee need is to be sure that structs used in OVMS and shared libraries are of same size.

Comment (dkalinowski) [reply src/custom_node_interface.hpp:29]: To make sure structs defined in OVMS and shared libraries are of the same size we need to compile it on the same architecture (x64 in this case)

File: src/node_library.cpp (hunk 3 above)
Comment (dkalinowski) [reply src/node_library.cpp]: Done

File: src/test/node_library_manager_test.cpp (hunk 5 above)
Comment (dkalinowski) [reply src/test/node_library_manager_test.cpp:48]: Done

File: src/custom_node_library_manager.cpp (hunk 2 above)
Comment (atobiszei) [reply src/custom_node_library_manager.cpp:59]: That's fair for retCode but log message to report what stage failed in logs would be usefull still. User can know that the library is missing execute/releaseBuffer or something else?

**Review by atobiszei — APPROVED** [review]

**Pull request merged**